weave

```diff
@@ -145,9 +170,11 @@ collect_ip_args() {
 if [ "$1" = "run" ] ; then
     shift 1
     check_docker_version
-    if [ "$1" = "--with-dns" ] ; then
-        shift 1
-        DNS_ARGS=$(exec_remote dns-args "$@")
+    collect_dns_args "$@"
+    shift $DNS_ARG_COUNT
+    [ $# -gt 0 ] || usage
+    if dns_available exec_remote dns-running ; then
+      DNS_ARGS=$(exec_remote dns-args "$@")
```

Comment:
We don't want to perform two `exec_remote`s here, especially not in the common case. Conversely, I wouldn't be too bothered if we *always* performed *one* exec_remote here, if that simplified matters.